src/List.mo

```diff
@@ -2891,4 +2891,47 @@ module {
     list.blockIndex == 1
   };
 
+  /// Unsafe iterator starting from `start`.
+  ///
+  /// Example:
+  /// ```
+  /// let list = List.fromArray<Nat>([1, 2, 3, 4, 5]);
+  /// let reader = List.reader<Nat>(list, 2);
+  /// assert reader() == 3;
+  /// assert reader() == 4;
+  /// assert reader() == 5;
+  /// ```
+  ///
+  /// Runtime: `O(1)`
+  ///
+  /// Space: `O(1)`
+  public func reader<T>(list : List<T>, start : Nat) : () -> T {
+    var blockIndex = 0;
+    var elementIndex = 0;
+    if (start != 0) {
+      let (block, element) = locate(start - 1);
+      blockIndex := block;
+      elementIndex := element + 1
+    };
+    var db : [var ?T] = list.blocks[blockIndex];
+    var dbSize = db.size();
+    func next() : T {
+      // Note: next() traps when reading beyond end of list
+      if (elementIndex == dbSize) {
+        blockIndex += 1;
+        db := list.blocks[blockIndex];
+        dbSize := db.size();
+        elementIndex := 0
+      };
+      switch (db[elementIndex]) {
+        case (?ret) {
+          elementIndex += 1;
+          return ret
+        };
+        case (_) Prim.trap(INTERNAL_ERROR)
```

Comment:
Maybe a different trap message to indicate that this an out-of-bounds access rather than an internal error:

```suggestion
        case (_) Prim.trap("List.reader(): out of bounds")
```

Note that `List.reader()` with the parentheses is a precedent set by other modules for the trap message.